### PR TITLE
[wmco] Run hybrid overlay as a Windows service

### PR DESCRIPTION
**Pull request by mansikulkarni96**
With the newly introduced feature, we can now run hybrid-overlay-node.exe as a Windows Service. This PR will leverage this functionality to start running hybrid-overlay-node as a Windows Service on the Windows nodes using WMCO.

### CODE REVIEW

File: pkg/controller/windowsmachine/windows/service.go [hunk 1]
```diff
@@ -17,30 +12,29 @@ type kubeProxyService struct {
...
+	return &service{
+		binaryPath: binaryPath,
+		name:       name,
+		args:       args,
 	}, nil
 }
 
 // Name returns the name of the service
-func (s *kubeProxyService) Name() string {
+func (s *service) Name() string {
```
Comment (LorbusChris) [inline pkg/controller/windowsmachine/windows/service.go:28]: if we go with a service struct, I think these methods can be dropped.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 2]
```diff
@@ -319,7 +324,7 @@ func (vm *windows) initializeBootstrapperFiles() error {
 }
 
 // createService creates the service on the Windows VM
-func (vm *windows) createService(svc service) error {
+func (vm *windows) createService(svc *service) error {
 	out, err := vm.Run("sc.exe create "+svc.Name()+" binPath=\""+svc.BinaryPath()+" "+
```
Comment (LorbusChris) [inline pkg/controller/windowsmachine/windows/windows.go:328]: instead of calling the methods here to return the fields, you could grab the struct fields contents here directly, i.e. `svc.name`, `svc.binaryPath`, and `svc.args`

**Review by LorbusChris — COMMENTED** [review]
looks good! I have just one small suggestion

File: pkg/controller/windowsmachine/windows/windows.go [hunk 3]
```diff
@@ -361,21 +366,6 @@ func (vm *windows) waitForHNSNetworks() error {
 	return errors.Wrap(err, "timeout waiting for OVN overlay HNS networks")
 }
 
-// waitForHybridOverlayToRun waits for the hybrid-overlay-node.exe to run until the timeout is reached
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:364]: Don't we need to wait for the service to go to running or do you think it waiting for HNS networks covers that? It will help us debug if we can be more specific regarding failures, i.e. the service did not go to running vs HNS networks did not get created.

Or does the service not go to running if there is an error in creating the HNS networks?

**Review by aravindhp — COMMENTED** [review]
Thanks for working on this, @mansikulkarni96

File: pkg/controller/windowsmachine/windows/windows.go (hunk 3 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:364]: I did think that waiting for HNS networks covers the scenario and we do not need to wait or retry more although I agree it would be easier to debug if we have that, AFAIK the code in hybrid-overlay-node the service will go to running state even with HNS network errors so we definitely need to waitForHNSNetworks

File: pkg/controller/windowsmachine/windows/service.go [hunk 4]
```diff
@@ -1,14 +1,9 @@
 package windows
 
-// service represents a Windows service
-type service interface {
-	Name() string
```
Comment (ravisantoshgudimetla) [inline pkg/controller/windowsmachine/windows/service.go:5]: Writing the comment here for record, @mansikulkarni96 you don't have to change the implementation now:

The downside of having just a struct for service instead of interface is you cannot have generic methods, for example, you've to expand windowsVM interface to include a method for getting status of a particular service. In a way, we're creating a dependency on the windowsVM to start and stop service. An ideal world would be something like:
```
type service interface {
Start()
Stop()
IsRunning()
Create()
Delete()
}

type windowsVM interface {
service
}
```

Now the flow would be a bit more natural in the sense, when we call vm.StartService(), we can directly call vm.Start()/ as the windows VM embeds service interface and all the methods are available in vm.StartService.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 5]
```diff
@@ -361,19 +369,20 @@ func (vm *windows) waitForHNSNetworks() error {
...
+// waitForHybridOverlayToRun waits for the hybrid-overlay-node service to reach RUNNING state
+// until the timeout is reached
 func (vm *windows) waitForHybridOverlayToRun() error {
 	var err error
 	for retries := 0; retries < retry.Count; retries++ {
-		_, err = vm.Run("Get-Process -Name \""+HybridOverlayProcess+"\"", true)
-		if err == nil {
+		out, _ := vm.Run("sc.exe query \""+hybridOverlayServiceName+"\" | findstr RUNNING", true)
```
Comment (ravisantoshgudimetla) [inline pkg/controller/windowsmachine/windows/windows.go:377]: What would happen if the service goes momentarily to running state and we return as success? Would it cause any problems or would it be caught as the HNS networks won't be created?

My main question is should we return when we see that our service goes to `RUNNING` or should we wait for few minutes and see `no failure`?

**Review by ravisantoshgudimetla — COMMENTED** [review]
Thank you for working on this @mansikulkarni96 

Just some minor questions.

File: pkg/controller/windowsmachine/windows/windows.go (hunk 5 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:377]: I am not sure if there can be a scenario like that AFAI can see from the code at: https://github.com/openshift/ovn-kubernetes/blob/2640a627e0a0c58d1d3fb71046f0bbf527522858/go-controller/hybrid-overlay/pkg/windows/service/service.go#L51, if there is an error service would not go to running state it will be returned before that and the only scenarios for service to go to start_pending after running are when the ctx is done or the service is stopped. To further add yes the HNSNetworks wont get created if there is an error after the service has been in running state. This function will just help us to debug more in the lines of whether the error was network related or service related. IMO we should be adding wait if its required.

File: pkg/controller/windowsmachine/windows/windows.go (hunk 3 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:364]: I have added this method back with changes required for checking is the service is running.

Comment (LorbusChris) [reply pkg/controller/windowsmachine/windows/windows.go:364]: Could we make this function more generic/re-usable by passing the service name as param? Something like:
```
func (vm *windows) waitForServiceToRun(serviceName string) error {
```
wdyt?

Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:364]: @LorbusChris that seems reasonable considering we could reuse this function for other services in the future. Thanks for the suggestion!

**Comment by LorbusChris** [comment]
thanks @mansikulkarni96!
/approve
/lgtm
(I'll leave the hold for another reviewer to lift)

File: pkg/controller/windowsmachine/windows/service.go [hunk 6]
```diff
@@ -17,30 +12,14 @@ type kubeProxyService struct {
...
-		binaryPath: kubeProxyPath,
-		name:       kubeProxyServiceName,
-		args: "--windows-service --v=4 --proxy-mode=kernelspace --feature-gates=WinOverlay=true " +
-			"--hostname-override=" + nodeName + " --kubeconfig=c:\\k\\kubeconfig " +
-			"--cluster-cidr=" + hostSubnet + " --log-dir=" + kubeProxyLogDir + " --logtostderr=false " +
-			"--network-name=OVNKubernetesHybridOverlayNetwork --source-vip=" + sourceVIP +
-			" --enable-dsr=false",
+// newService initializes and returns a pointer to the service struct
+func newService(binaryPath, name, args string) (*service, error) {
+	if args == "" || binaryPath == "" || name == "" {
```
Comment (sebsoto) [inline pkg/controller/windowsmachine/windows/service.go:17]: Args should be able to be empty, as a Windows service being called with no arguments is a valid case

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/controller/windowsmachine/windows/service.go [hunk 7]
```diff
@@ -17,30 +12,14 @@ type kubeProxyService struct {
...
-		name:       kubeProxyServiceName,
-		args: "--windows-service --v=4 --proxy-mode=kernelspace --feature-gates=WinOverlay=true " +
-			"--hostname-override=" + nodeName + " --kubeconfig=c:\\k\\kubeconfig " +
-			"--cluster-cidr=" + hostSubnet + " --log-dir=" + kubeProxyLogDir + " --logtostderr=false " +
-			"--network-name=OVNKubernetesHybridOverlayNetwork --source-vip=" + sourceVIP +
-			" --enable-dsr=false",
+// newService initializes and returns a pointer to the service struct
+func newService(binaryPath, name, args string) (*service, error) {
+	if binaryPath == "" || name == "" {
+		return nil, errors.Errorf("can't instantiate a service with incomplete service arguments")
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/service.go:18]: The error seems to imply `args` is incomplete

File: pkg/controller/windowsmachine/windows/windows.go [hunk 8]
```diff
@@ -47,6 +47,10 @@ const (
 	OVNKubeOverlayNetwork = "OVNKubernetesHybridOverlayNetwork"
 	// kubeProxyServiceName is the name of the kube-proxy Windows service
 	kubeProxyServiceName = "kube-proxy"
+	// kubeletServiceName is the name of the kubelet Windows service
+	kubeletServiceName = "kubelet"
+	// hybridOverlayServiceName is the name of the hybrid-overlay-node Windows service
+	hybridOverlayServiceName = "hybrid-overlay-node"
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:53]: The hybrid-verlay related items were in one spot. Any reason to move it?

File: pkg/controller/windowsmachine/windows/windows.go [hunk 9]
```diff
@@ -161,31 +165,41 @@ func (vm *windows) Configure() error {
...
-	// err being nil implies that hybrid-overlay is running.
-	if err == nil {
-		// Stop the hybrid-overlay
-		stopCmd := "Stop-Process -Name \"" + HybridOverlayProcess + "\""
-		out, err := vm.Run(stopCmd, true)
-		if err != nil {
-			log.Info("unable to stop hybrid-overlay", "stop command", stopCmd, "output", out)
-			return errors.Wrap(err, "unable to stop hybrid-overlay")
-		}
+	// Ensure kubelet service is running before configuring hybrid-overlay-node Windows service
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:168]: I would think this would be expressed using a dependency.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 10]
```diff
@@ -231,9 +231,16 @@ func (vm *windows) ConfigureKubeProxy(nodeName, hostSubnet string) error {
 	if err != nil {
 		return errors.Wrap(err, "error getting source VIP")
 	}
-	kubeProxyService, err := newKubeProxyService(nodeName, hostSubnet, sVIP)
+
+	kubeProxyServiceArgs := "--windows-service --v=4 --proxy-mode=kernelspace --feature-gates=WinOverlay=true " +
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:249]: Doesn't kube-proxy depend on hybrid-overlay service?

File: pkg/controller/windowsmachine/windows/windows.go [hunk 11]
```diff
@@ -236,7 +250,7 @@ func (vm *windows) ConfigureKubeProxy(nodeName, hostSubnet string) error {
 		"--hostname-override=" + nodeName + " --kubeconfig=c:\\k\\kubeconfig " +
 		"--cluster-cidr=" + hostSubnet + " --log-dir=" + kubeProxyLogDir + " --logtostderr=false " +
 		"--network-name=OVNKubernetesHybridOverlayNetwork --source-vip=" + sVIP +
-		" --enable-dsr=false"
+		" --enable-dsr=false\""
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:253]: Why is this change needed? If at it is, then it needs to be in the previous commit.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 12]
```diff
@@ -348,27 +362,32 @@ func (vm *windows) initializeBootstrapperFiles() error {
 // createService creates the service on the Windows VM
 func (vm *windows) createService(svc *service) error {
 	out, err := vm.Run("sc.exe create "+svc.name+" binPath=\""+svc.binaryPath+" "+
-		svc.args+"\" start=auto", false)
+		svc.args+" start=auto", false)
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:365]: Shouldn't this be in the restructuring commit?

File: pkg/controller/windowsmachine/windows/windows.go [hunk 13]
```diff
@@ -348,27 +362,32 @@ func (vm *windows) initializeBootstrapperFiles() error {
...
 }
 
 // stopService stops the service that was already running
-func (vm *windows) stopService(svc service) error {
-	out, err := vm.Run("sc.exe stop "+svc.Name(), false)
+func (vm *windows) stopService(svc *service) error {
+	out, err := vm.Run("sc.exe stop "+svc.name, false)
 	if err != nil {
-		return errors.Wrapf(err, "failed to stop %s service with output: %s", svc.Name(), out)
+		return errors.Wrapf(err, "failed to stop %s service with output: %s", svc.name, out)
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:376]: Shouldn't this be in the restructuring commit?

File: pkg/controller/windowsmachine/windows/windows.go [hunk 14]
```diff
@@ -348,27 +362,32 @@ func (vm *windows) initializeBootstrapperFiles() error {
...
 	}
 	return nil
 }
 
 // isRunning gets the status of given service
-func (vm *windows) isRunning(svc service) bool {
-	out, err := vm.Run("sc.exe query "+svc.Name(), false)
+func (vm *windows) isRunning(svc *service) bool {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:382]: This function now does two things. Moreover it is returning `false` when the service does not exists which implies the service exists but is not running. This is incorrect semantics.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 15]
```diff
@@ -378,12 +397,22 @@ func (vm *windows) isRunning(svc service) bool {
...
 	out, err := vm.Run("sc.exe start "+svc.name, false)
 	if err != nil {
-		return errors.Wrapf(err, "failed to start %s service with output: %s", svc.Name(), out)
+		return errors.Wrapf(err, "failed to start %s service with output: %s", svc.name, out)
 	}
 	log.V(1).Info("started service", "name", svc.name, "binary", svc.binaryPath, "args", svc.args)
 	return nil
 }
 
+//queryServiceConfig queries a service for it's configuration
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:406]: Space after `//`

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @mansikulkarni96. Please address my comments. I will leave the service struct restructuring review to @ravisantoshgudimetla and @sebsoto.

File: pkg/controller/windowsmachine/windows/service.go (hunk 7 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/service.go]: edited msg

File: pkg/controller/windowsmachine/windows/windows.go (hunk 8 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go]: moved it to the top

File: pkg/controller/windowsmachine/windows/windows.go (hunk 9 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go]: yes removed this

File: pkg/controller/windowsmachine/windows/windows.go (hunk 10 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:247]: added dependency

File: pkg/controller/windowsmachine/windows/windows.go (hunk 11 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go]: needed as the depend args are being added, moved to previous commit.

File: pkg/controller/windowsmachine/windows/windows.go (hunk 14 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go]: split into two, that checks for existing first then checks if running.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 16]
```diff
@@ -418,19 +430,19 @@ func (vm *windows) waitForHNSNetworks() error {
 	return errors.Wrap(err, "timeout waiting for OVN overlay HNS networks")
 }
 
-// waitForHybridOverlayToRun waits for the hybrid-overlay-node.exe to run until the timeout is reached
-func (vm *windows) waitForHybridOverlayToRun() error {
+// waitForServiceToRun waits for the given service to be in RUNNING state
+// until the timeout is reached
+func (vm *windows) waitForServiceToRun(serviceName string) error {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:435]: Shouldn't this be in the service restructuring commit?

File: pkg/controller/windowsmachine/windows/windows.go [hunk 17]
```diff
@@ -161,31 +165,39 @@ func (vm *windows) Configure() error {
...
 
+	// if the service exits and it is running, then stop it.
+	if vm.serviceExists(hybridOverlayServiceName) {
+		if vm.isRunning(hybridOverlayServiceName) {
+			if err := vm.stopService(hybridOverlayService); err != nil {
+				return errors.Wrapf(err, "error stopping %s Windows service", hybridOverlayServiceName)
+			}
+		}
+	} else {
+		if err := vm.createService(hybridOverlayService); err != nil {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:189]: Can't this be combined? If yes, please do that for the kube-proxy case too.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/controller/windowsmachine/windows/windows.go (hunk 16 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:435]: I have changed this method to accomodate for the service that I am adding, hence it made  sense to me to keep it in the feature commit.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 18]
```diff
@@ -339,41 +360,50 @@ func (vm *windows) initializeBootstrapperFiles() error {
...
+		return errors.Wrapf(err, "failed to stop %s service with output: %s", svc.name, out)
 	}
 	return nil
 }
 
+// serviceExists check if the given service exists on Windows VM
+func (vm *windows) serviceExists(serviceName string) bool {
+	_, err := vm.Run("sc.exe qc "+serviceName, false)
+	if err != nil {
+		return false
```
Comment (sebsoto) [inline pkg/controller/windowsmachine/windows/windows.go:385]: You need to check if the error was a code 1060 and only then return false, else return error.

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/controller/windowsmachine/windows/windows.go (hunk 18 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:385]: I thought that was necessary because we weren't checking for service existence before querying the service for running status and that was causing the reconcile errors, hence we want to check the error code it is throwing. This is an alternative approach where I am using qc to check if service exists and query to check for running status. I can log the error here before returning false.

Comment (sebsoto) [reply pkg/controller/windowsmachine/windows/windows.go:385]: its important that we make the distinction between `sc.exe qc` returning error from the service not existing, and it returning error because of various other reasons. If the service exists and we have a transient ssh error, we will think the service doesnt exist when it actually does.

Comment (aravindhp) [reply pkg/controller/windowsmachine/windows/windows.go:385]: @mansikulkarni96 If you go down this route we have to be very careful about parsing the errors. Just looking for `1060` is not sufficient. Then you need to return an `error` in addition to the `bool`.

File: pkg/controller/windowsmachine/windows/windows.go (hunk 17 above)
Comment (aravindhp) [reply pkg/controller/windowsmachine/windows/windows.go:189]: Please note the comment is about two lines, 188 to 189. I am asking if this can be made into an `if else`. But thinking about your question over slack, I think lines 182-192 can be made into a function like `ensureServiceExists()` and used for both kube-proxy and hybrid-overlay services.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 19]
```diff
@@ -339,41 +348,59 @@ func (vm *windows) initializeBootstrapperFiles() error {
 }
 
 // createService creates the service on the Windows VM
-func (vm *windows) createService(svc service) error {
-	out, err := vm.Run("sc.exe create "+svc.Name()+" binPath=\""+svc.BinaryPath()+" "+
-		svc.Args()+"\" start=auto", false)
+func (vm *windows) createService(svc *service) error {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:361]: Nit: ideally all functions that take a pointer input should have a nil pointer check.

**Review by aravindhp — COMMENTED** [review]
LGTM

File: pkg/controller/windowsmachine/windows/windows.go [hunk 20]
```diff
@@ -339,41 +358,59 @@ func (vm *windows) initializeBootstrapperFiles() error {
...
-// isRunning gets the status of given service
-func (vm *windows) isRunning(svc service) bool {
-	out, err := vm.Run("sc.exe query "+svc.Name(), false)
+// serviceExists checks if the given service exists on Windows VM
+func (vm *windows) serviceExists(serviceName string) (bool, error) {
+	_, err := vm.Run("sc.exe qc "+serviceName, false)
 	if err != nil {
-		log.Info("failed to query service", "service", svc.Name(), "output", out, "err", err)
-		return false
+		if strings.Contains(err.Error(), "status 1060") {
```
Comment (sebsoto) [inline pkg/controller/windowsmachine/windows/windows.go:383]: this could use a comment explaining the 1060 error code

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/controller/windowsmachine/windows/windows.go (hunk 20 above)
Comment (aravindhp) [reply pkg/controller/windowsmachine/windows/windows.go:383]: Good point, @sebsoto. @mansikulkarni96 please link to https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--1000-1299-

File: pkg/controller/windowsmachine/windows/windows.go [hunk 21]
```diff
@@ -339,41 +358,59 @@ func (vm *windows) initializeBootstrapperFiles() error {
...
+		return false, err
+	}
+	return strings.Contains(out, "RUNNING"), nil
 }
 
 // startService starts a previously created Windows service
-func (vm *windows) startService(svc service) error {
-	out, err := vm.Run("sc.exe start "+svc.Name(), false)
+func (vm *windows) startService(svc *service) error {
+	serviceRunning, err := vm.isRunning(svc.name)
```
Comment (ravisantoshgudimetla) [inline pkg/controller/windowsmachine/windows/windows.go:402]: nit: Checking for the current state should be responsibility of caller method.

**Review by ravisantoshgudimetla — APPROVED** [review]
/approve


Thank you for working on this @mansikulkarni96 
with the caveat that you will address other's comments

File: pkg/controller/windowsmachine/windows/windows.go (hunk 21 above)
Comment (mansikulkarni96) [reply pkg/controller/windowsmachine/windows/windows.go:413]: as discussed in the discussion earlier I am avoiding this here to avoid code repetition in both service caller methods.

**Review by sebsoto — APPROVED** [review]
/lgtm

**Comment by LorbusChris** [comment]
:shipit: 

/lgtm
/hold cancel
/retest

**Comment by aravindhp** [comment]
@LorbusChris, we typically wait for the PR author to remove the hold as they potentially could have more changes in mind :-)
